Preserve block style variations when securing theme json

dsas opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Preserves global styles block variations when securing styles.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Global Styles block variations were being removed for users whose capabilities dictated that their styles changes needed securing, despite not actually being insecure. When this was a problem varied depending upon site configuration, but out-of-the-box it was a problem for administrators on multi-site installs.

Valid and safe block style variations were being removed by `WP_Theme_JSON_Gutenberg::remove_insecure_properties` when securing styles. 

Closes https://github.com/WordPress/gutenberg/issues/53428

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Adds explicit processing of variations in `remove_insecure_properties` so that they are preserved not removed.

## Testing Instructions

1. Activate latest Gutenberg on a multisite install, or test with a user not having the unfiltered_html capability, or enable the KSES filters with add_action( 'init', 'kses_init_filters' );
2. Go to Appearance > Editor
3. Open the Styles sidebar
4. Select "Blocks"
5. Select "Button"
6. Select the Outline style variation
7. Change a property away from the initial setting, e.g. set the background
8. Save the changes
9.  ⚠️ Note how the changes are reverted

Apply this PR and try again - note how the changes are now not reverted.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
N/A

## Screenshots or screencast <!-- if applicable -->

dsas committed:
Preserve block style variations when securing theme json

Valid and safe block style variations were being removed by
`WP_Theme_JSON_Gutenberg::remove_insecure_properties` when securing the
theme.json. When this was a problem varied depending upon site
configuration, but out-of-the-box it was a problem for administrators on
multi-site installs.

This change adds explicit processing of variations in
`remove_insecure_properties` so that they won't get removed.

andrewserong commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -1702,6 +1702,34 @@ public function data_get_styles_for_block_with_style_variations() {
+						'variations' => array(
+							'outline' => array(
+								'color' => array(
+									'background' => 'purple',
```
Not a blocker, but to be extra thorough about the test, would it be worth adding an invalid property to the `outline` array, to test that it correctly gets removed by the call to `remove_insecure_styles`?

dsas replied:
Cheers @andrewserong I've added another test for that

andrewserong reviewed (approved):
This is testing great for me, thanks for following up with the PR and for the thorough testing instructions!

✅ Confirmed the issue on trunk when running without `unfiltered_html` capability
✅ Confirmed this fixes the issue and I'm able to save styling changes to a variation within the site editor
✅ Logic change looks good to me and is consistent with the rest of the code in `remove_insecure_properties`  👍

Just left an optional comment for the test about whether we should also check that invalid properties for the variation are stripped, but otherwise this LGTM! ✨

dsas committed:
Add another variation sanitisation test

This test checks that when removing insecure properties an
unknown/unsupported property is removed from the variation.

andrewserong reviewed (approved):
Thanks for the update, the additional test covers things nicely! LGTM, I'll merge this in now 🙂

tellthemachines commented:
I just cherry-picked this PR to the update/bugfixes-6-3-1 branch to get it included in the next release: ee2786ad9f

ramonjd commented:
Hi @dsas 

These changes will need to be synced with Core. Are you okay to do that? No worries if not. Let us know and I can take care of it.

If you want to take it on, it is recommended to create a [new Trac ticket](https://core.trac.wordpress.org/newticket) and submit a pull request to the [WordPress Core Github repository](https://github.com/WordPress/wordpress-develop).

Thank you! :heart:

dsas commented:
Thanks @ramonjd I've opened https://github.com/WordPress/wordpress-develop/pull/5013, I haven't done it before so I hope I've got everything right!
